plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/conversion/StmtConversionVisitor.kt

```diff
@@ -235,12 +234,46 @@ object StmtConversionVisitor : FirVisitor<Exp, StmtConversionContext<ResultTrack
         data.addStatement(Stmt.If(Exp.BoolLit(true), inlineCtx.block, Stmt.Seqn(listOf(), listOf())))
     }
 
+    private fun processInlineLambdaArgumentExpression(
+        lambdaArgumentExpression: FirLambdaArgumentExpression,
+        data: StmtConversionContext<ResultTrackingContext>
+    ): Stmt.Seqn =
+        when (val expr = lambdaArgumentExpression.expression) {
+            is FirAnonymousFunctionExpression -> {
+                val body = expr.anonymousFunction.body!!
+                data.convert(body)
+                data.block
+            }
+            else -> TODO("are there any other cases?")
+        }
+
     override fun visitImplicitInvokeCall(
         implicitInvokeCall: FirImplicitInvokeCall,
         data: StmtConversionContext<ResultTrackingContext>,
     ): Exp {
-        val args = getFunctionCallArguments(implicitInvokeCall).map(data::convert)
         val retType = implicitInvokeCall.calleeCallableSymbol.resolvedReturnType
+        val calleeName = LocalName(implicitInvokeCall.calleeReference.name)
+        val lambda = data.getLambdaArg(calleeName)
+        if (lambda != null) {
+            return data.withResult(data.embedType(retType)) {
+                val lambdaExpr = (lambda.expression as FirAnonymousFunctionExpression).anonymousFunction
+                // NOTE: it is not needed to make distinction between implicit or explicit parameters
+                val lambdaParams = lambdaExpr.valueParameters.map { LocalName(it.name) }
+                val callArgs = getFunctionCallSubstitutionItems(implicitInvokeCall, data)
+                // Note: The first element of callArgs is the callee and has to be dropped
+                val subs: Map<MangledName, SubstitutionItem> = lambdaParams.zip(callArgs.drop(1)).toMap()
+                val sqn = processInlineLambdaArgumentExpression(
+                    lambda,
+                    this.newBlock().withInlineContext(this.signature, this.resultCtx.resultVar, subs, false)
+                )
+                // TODO: A wrong goto is added at the end of each lambda, for now I just remove it
+                val stmtsWithoutGoTo = sqn.stmts.dropLast(1)
```

Comment:
This seems like a problem with the signature; it should be the parent context signature that gets passed here.  That said, the `return` should refer to the label of the lambda, not the one of the function.  I agree it gets a bit messy.  Perhaps this information can be stashed away when the lambda is constructed?